Optimize functions with useCallback for improved performance and memory usage 🚀

sanjaiyan-dev opened the pull request:
# Why
This pull request addresses performance and memory usage by wrapping specific functions with `useCallback`. This optimization technique leverages memoization, ensuring that these functions are only recreated when their dependencies change. This reduces unnecessary re-renders and minimizes memory allocation, leading to improved overall application performance.

Ref-: https://react.dev/reference/react/useCallback#skipping-re-rendering-of-components


Extremely sorry if I made any mistakes :)

# Checklist

<!--
Please check the appropriate items below if they apply to your diff.
-->

- [ ] I added a `changelog.md` entry and rebuilt the package sources according to [this short guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

sanjaiyan-dev committed:
Optimize canvas.tsx rocket🚀

sanjaiyan-dev committed:
Merge branch 'expo:main' into sanjaiyan-gl-optimize

lukmccall commented on packages/expo-gl/src/Canvas.tsx:
```diff
@@ -37,7 +37,18 @@ const CanvasWrapper: React.FunctionComponent<
+    const element = getElement(ref.current);
+    const { offsetWidth: width = 0, offsetHeight: height = 0 } = element;
+    return { width, height };
+  }, [!!size, ref.current])
```
Was there any reason for converting the size to a boolean value?

sanjaiyan-dev replied:
Hello! I had a thought about the `size` object. Since JavaScript compares objects by reference, converting it to a Boolean for comparison might be more stable and performant, especially since the function seems to only check for the presence of the `size` object, not its specific values. Alternatively, we could directly compare `size.width` and `size.height` if we need to examine the dimensions.

I apologize if my understanding is incorrect.

sanjaiyan-dev replied:
> Was there any reason for converting the size to a boolean value?

https://github.com/expo/expo/pull/35601#issuecomment-2800139074

sanjaiyan-dev committed:
Update Canvas.tsx 🎨

Fix minor bugs

Remove mutable reference from react dependency array and update react dependencies to be more predictable.

sanjaiyan-dev commented:
Apologies for the delayed response.

Regarding this comment: [https://github.com/facebook/react/issues/14387#issuecomment-503616820](https://github.com/facebook/react/issues/14387#issuecomment-503616820)

I realise now that including React references in the dependency array is not good practice. Sorry for the earlier error – I've now updated the dependency array.

lukmccall commented on packages/expo-gl/src/Canvas.tsx:
```diff
@@ -37,7 +37,18 @@ const CanvasWrapper: React.FunctionComponent<
   const _canvasRef = React.useRef<HTMLCanvasElement>(null);
 
-  function updateCanvasSize(): void {
+  const getSize = React.useCallback((): { width: number; height: number } => {
+    if (size) {
+      return size;
+    } else if (!ref.current || typeof window === 'undefined') {
+      return { width: 0, height: 0 };
+    }
+    const element = getElement(ref.current);
+    const { offsetWidth: width = 0, offsetHeight: height = 0 } = element;
+    return { width, height };
+  }, [size?.width, size?.height])
```
It would be better to move it to a top level and pass needed arguments as parameters (size and ref.current). It will make the code more readable, now it's unclear what is the relation between `getSize` and `onLayout`.

lukmccall replied:
Indeed, we would need to pass the current size to the function, but I believe this approach will improve readability and eliminate the need to track dependencies. The `onLayout` function must remain within the component since it modifies the current state. I was only referring to the `getSize` function.

sanjaiyan-dev replied:
> Indeed, we would need to pass the current size to the function, but I believe this approach will improve readability and eliminate the need to track dependencies. The `onLayout` function must remain within the component since it modifies the current state. I was only referring to the `getSize` function.

Hi, I needed to create a new PR (https://github.com/expo/expo/pull/36214) because of ongoing merge conflict issues. Please use this one going forward. Thanks!

lukmccall commented:
Closing in favor of https://github.com/expo/expo/pull/36214.
